In [2]:
import sys
sys.path.append('../source/')
import numpy as np
from Users import Users
from Producers import *
from embeddings import *
from sklearn.preprocessing import normalize
from plotting import *
from tqdm import tqdm
import pandas as pd

In [6]:
utils_columns = ['total_prod_util','avg_prod_util', 'max_prod_util', 'min_prod_util', \
                 'total_user_util','avg_user_util','max_user_util', 'min_user_util']

### Uniform synthetic
- 399 instances conveged to NE for linear, all 400 converged for softmax
- Of those 399 instances, Linear is better than softmax for both producer and user utilities (except for 2 instances for the max_user_util column)

In [22]:
df_linear = pd.read_pickle("../saved_frames/uniform_synthetic_grid_linear_statistics.pkl") 
df_softmax = pd.read_pickle("../saved_frames/uniform_synthetic_grid_softmax_statistics.pkl") 

In [37]:
df_linear.shape, df_softmax.shape, df_linear.columns

((400, 16),
 (400, 16),
 Index(['dimension', 'seed', 'nprod', 'NE_exists', 'iters',
        'last_profile_compact', 'producer_dist', 'user_dist', 'total_prod_util',
        'avg_prod_util', 'max_prod_util', 'min_prod_util', 'total_user_util',
        'avg_user_util', 'max_user_util', 'min_user_util'],
       dtype='object'))

In [24]:
df_linear['NE_exists'].sum(), df_softmax['NE_exists'].sum()

(399, 400)

In [27]:
mask = df_linear['NE_exists']
(df_softmax[mask][utils_columns] < df_linear[mask][utils_columns]).sum()

total_prod_util    399
avg_prod_util      399
max_prod_util      399
min_prod_util      399
total_user_util    399
avg_user_util      399
max_user_util      397
min_user_util      399
dtype: int64

In [30]:
# difference, linear - softmax for utilities
(df_linear[mask][utils_columns] - df_softmax[mask][utils_columns]).describe()

total_prod_util  avg_prod_util  max_prod_util  min_prod_util  \
count       399.000000     399.000000     399.000000     399.000000   
mean        684.369551      19.842705      20.836021      18.974458   
std         237.304718      20.078576      21.233363      19.019626   
min         388.680356       4.222688       4.302907       4.136017   
25%         465.475908       7.909494       8.108752       7.760211   
50%         718.950983      12.316220      12.754441      12.065529   
75%         886.795244      21.364928      23.412850      20.499442   
max        1060.819757     106.081976     114.069056     102.894447   

       total_user_util  avg_user_util  max_user_util  min_user_util  
count       399.000000     399.000000     399.000000     399.000000  
mean        684.369551       0.068437       0.244095       0.049992  
std         237.304718       0.023730       0.148684       0.024368  
min         388.680356       0.038868      -0.002132       0.000484  
25%         465.475908       0.046548       0.128423       0.027502  
50%         718.950983       0.071895       0.191765       0.054302  
75%         886.795244       0.088680       0.393687       0.067307  
max        1060.819757       0.106082       0.515449       0.090755

In [43]:
df_linear[(df_linear['dimension'] == 5) & (df_linear['nprod'] == 10)]

dimension  seed  nprod  NE_exists  iters       last_profile_compact  \
0           5     2     10       True      3  [2.0, 2.0, 2.0, 2.0, 2.0]   
10          5     3     10       True      3  [2.0, 2.0, 2.0, 2.0, 2.0]   
20          5     5     10       True      3  [2.0, 2.0, 2.0, 2.0, 2.0]   
30          5     7     10       True      4  [2.0, 2.0, 2.0, 2.0, 2.0]   
40          5    11     10       True      2  [2.0, 2.0, 2.0, 2.0, 2.0]   
50          5    13     10       True      5  [2.0, 2.0, 2.0, 2.0, 2.0]   
60          5    17     10       True      1  [2.0, 2.0, 2.0, 2.0, 2.0]   
70          5    19     10       True      3  [2.0, 2.0, 2.0, 2.0, 2.0]   
80          5    23     10       True      6  [2.0, 2.0, 2.0, 2.0, 2.0]   
90          5    29     10       True      3  [2.0, 2.0, 2.0, 2.0, 2.0]   

                producer_dist  \
0   [0.2, 0.2, 0.2, 0.2, 0.2]   
10  [0.2, 0.2, 0.2, 0.2, 0.2]   
20  [0.2, 0.2, 0.2, 0.2, 0.2]   
30  [0.2, 0.2, 0.2, 0.2, 0.2]   
40  [0.2, 0.2, 0.2, 0.2, 0.2]   
50  [0.2, 0.2, 0.2, 0.2, 0.2]   
60  [0.2, 0.2, 0.2, 0.2, 0.2]   
70  [0.2, 0.2, 0.2, 0.2, 0.2]   
80  [0.2, 0.2, 0.2, 0.2, 0.2]   
90  [0.2, 0.2, 0.2, 0.2, 0.2]   

                                            user_dist  total_prod_util  \
0   [0.1985065740914945, 0.20296485834765995, 0.19...      3348.699486   
10  [0.19963523503021607, 0.2007249606838913, 0.20...      3328.041807   
20  [0.20257287206198313, 0.2012798194440244, 0.19...      3320.530466   
30  [0.19596202585469547, 0.20099394506356583, 0.2...      3336.911084   
40  [0.20113254531133956, 0.19795997909784113, 0.2...      3337.288030   
50  [0.20007785926324337, 0.2000912695958865, 0.20...      3329.396009   
60  [0.19795098440854125, 0.20069698840386346, 0.1...      3342.700987   
70  [0.19734045561066657, 0.19983119774218225, 0.2...      3329.666701   
80  [0.19886916795502832, 0.1982992018850868, 0.20...      3337.953092   
90  [0.20068460716651632, 0.20311781841797288, 0.1...      3336.767143   

    avg_prod_util  max_prod_util  min_prod_util  total_user_util  \
0      334.869949     340.204491     329.684015      3348.699486   
10     332.804181     335.949562     330.103021      3328.041807   
20     332.053047     339.197666     323.747447      3320.530466   
30     333.691108     344.924072     320.211504      3336.911084   
40     333.728803     339.315320     327.654683      3337.288030   
50     332.939601     336.282101     328.865792      3329.396009   
60     334.270099     344.894542     328.268123      3342.700987   
70     332.966670     341.402426     324.000152      3329.666701   
80     333.795309     341.597091     327.985219      3337.953092   
90     333.676714     342.335211     327.173974      3336.767143   

    avg_user_util  max_user_util  min_user_util  
0        0.334870       0.918208       0.201381  
10       0.332804       0.811231       0.200605  
20       0.332053       0.899242       0.200694  
30       0.333691       0.885513       0.201170  
40       0.333729       0.936052       0.201612  
50       0.332940       0.889383       0.200905  
60       0.334270       0.876490       0.201201  
70       0.332967       0.929049       0.200844  
80       0.333795       0.879303       0.201226  
90       0.333677       0.897498       0.200718

### Sorted synthetic
- all 400 instances converged to NE
- Softmax is better for producer and user utility in majority of cases

In [44]:
df_linear = pd.read_pickle("../saved_frames/sorted_synthetic_grid_linear_statistics.pkl") 
df_softmax = pd.read_pickle("../saved_frames/sorted_synthetic_grid_softmax_statistics.pkl") 

In [45]:
df_linear['NE_exists'].sum(), df_softmax['NE_exists'].sum()

(400, 400)

In [46]:
mask = df_linear['NE_exists']
(df_softmax[mask][utils_columns] < df_linear[mask][utils_columns]).sum()

total_prod_util     96
avg_prod_util       95
max_prod_util      119
min_prod_util      119
total_user_util     36
avg_user_util       33
max_user_util      110
min_user_util      109
dtype: int64

In [48]:
(df_softmax[mask][utils_columns] < df_linear[mask][utils_columns])

total_prod_util  avg_prod_util  max_prod_util  min_prod_util  \
0              False          False          False          False   
1              False          False          False          False   
2              False          False          False          False   
3              False          False          False          False   
4              False          False          False          False   
..               ...            ...            ...            ...   
395             True           True           True           True   
396            False          False           True           True   
397            False          False          False          False   
398            False          False           True           True   
399             True           True           True           True   

     total_user_util  avg_user_util  max_user_util  min_user_util  
0              False          False           True          False  
1              False          False          False          False  
2              False          False          False           True  
3              False          False          False          False  
4              False          False          False           True  
..               ...            ...            ...            ...  
395            False          False           True          False  
396            False          False          False          False  
397            False          False           True          False  
398            False          False          False          False  
399            False          False          False          False  

[400 rows x 8 columns]

### Movielens-100k
- 398 converged for linear to NE, all 400 instances converged for softmax to NE
- linear is better for utility of producer and consumer in majority of cases

In [49]:
df_linear = pd.read_pickle("../saved_frames/Movielens100k_grid_linear_statistics.pkl") 
df_softmax = pd.read_pickle("../saved_frames/Movielens100k_grid_softmax_statistics.pkl") 

In [50]:
df_linear['NE_exists'].sum(), df_softmax['NE_exists'].sum()

(398, 400)

In [51]:
mask = df_linear['NE_exists']
(df_softmax[mask][utils_columns] < df_linear[mask][utils_columns]).sum()

total_prod_util    398
avg_prod_util      398
max_prod_util      398
min_prod_util      398
total_user_util    398
avg_user_util      398
max_user_util      237
min_user_util      398
dtype: int64

In [52]:
(df_softmax[mask][utils_columns] < df_linear[mask][utils_columns])

total_prod_util  avg_prod_util  max_prod_util  min_prod_util  \
0               True           True           True           True   
1               True           True           True           True   
2               True           True           True           True   
3               True           True           True           True   
4               True           True           True           True   
..               ...            ...            ...            ...   
395             True           True           True           True   
396             True           True           True           True   
397             True           True           True           True   
398             True           True           True           True   
399             True           True           True           True   

     total_user_util  avg_user_util  max_user_util  min_user_util  
0               True           True           True           True  
1               True           True           True           True  
2               True           True           True           True  
3               True           True           True           True  
4               True           True           True           True  
..               ...            ...            ...            ...  
395             True           True          False           True  
396             True           True          False           True  
397             True           True          False           True  
398             True           True          False           True  
399             True           True          False           True  

[398 rows x 8 columns]